docs/adr/0012_construct_as_transformation.md

```diff
@@ -0,0 +1,441 @@
+    #     data: <binary data>
+    
+  - type: resource.downloader/v1alpha1 # maps to the access or input type
+    id: downloadresource2
+    resource: ${resourcecreator2.resource}
+    # output:
+    #   resource:
+    #     name: localociartifact
+    #     type: blob
+    #     relation: local
+    #     access:
+    #       type: ociArtifact
+    #       imageReference: ghcr.io/fabianburth/artifact:1.0.0
+    #     data: <binary data>
+    
+    # NO DOWNLOADER NEEDED FOR downloadresource3, WE WANT TO ADD IT BY REFERENCE
+    
+  - type: local.resource.uploader/v1alpha1
+    id: uploadresource1
+    resource: ${downloadresource1.resource}
+    # the repository and the component have to match with the component 
+    # to which the resource is added and the repository where the component 
+    # is uploaded
+    repository: ${helloworld.repositorySpec}
+    component: ${helloworld.componentIdentity}
+    # output:
+    #   resource:
+    #     name: localtext
+    #     type: blob
+    #     relation: local
+    #     access:
+    #       type: localblob
+    #       localReference: <reference in target repo>
+
+  - type: local.resource.uploader/v1alpha1
+    id: uploadresource2
+    resource: ${downloadresource2.resource}
+    repository: ${attributes.repositorySpec}
+    component: ${attributes.componentIdentity}
+    # output:
+    #   resource:
+    #     name: localociartifact
+    #     type: blob
+    #     relation: local
+    #     access:
+    #       type: localblob
+    #       localReference: <reference in target repo>
+
+
+  - type: attribute
+    id: helloworld-ref
+    componentIdentity:
+      name: github.com/acme.org/helloworld-ref
+      version: 1.0.0
+    repositorySpec:
+      type: ociRegistry
+      baseUrl: "ghcr.io"
+      subPath: "/open-component-model/target-ocm-repository"
+
+  - type: component.download
+    id: downloadcomponentversion2
+    component: ${helloworld-ref.componentIdentity}
+    # output: < component descriptor >
+  
+  - type: component.digest/v1alpha1
+    id: digestcomponentversion2
+    componentDescriptor: ${downloadcomponentversion2.output}
+    # output:
+    #   digest: sha256:<hash>
+    
+  - type: attribute
+    id: createcomponentversion1
+    name: ${helloworld.componentIdentity.name} 
+    version: ${helloworld.componentIdentity.version}
+    provider:
+      name: internal
+    resources:
+      - ${uploadresource1.resource}
+      - ${uploadresource2.resource}
+      - ${resourcecreator3.resource} # add by reference, no upload needed
+    componentReferences:
+      - name: ref
+        version: ${downloadcomponentversion2.output.version}
+        componentName: ${downloadcomponentversion2.output.name}
+        digest: ${digestcomponentversion2.output.digest}
+
+  - type: component.uploader/v1alpha1
+    id: uploadcomponentversion1
+    repository: ${attributes.repositorySpec}
+    componentDescriptor: ${createcomponentversion1.outputs.descriptor}
+```
```

Comment:
I would use a smaller example here because it's rather large and easy to get lost. Use more specific ones and smaller ones further down the line to make a point?